app/src/main/java/org/simple/clinic/pin/PinEntryCardController.kt

```diff
@@ -0,0 +1,81 @@
+package org.simple.clinic.pin
+
+import io.reactivex.Observable
+import io.reactivex.ObservableSource
+import io.reactivex.ObservableTransformer
+import io.reactivex.rxkotlin.ofType
+import org.simple.clinic.login.applock.ComparisonResult
+import org.simple.clinic.login.applock.ComparisonResult.DIFFERENT
+import org.simple.clinic.login.applock.ComparisonResult.SAME
+import org.simple.clinic.login.applock.PasswordHasher
+import org.simple.clinic.pin.PinEntryCardView.State
+import org.simple.clinic.user.UserSession
+import org.simple.clinic.widgets.UiEvent
+import javax.inject.Inject
+
+typealias Ui = PinEntryCardView
+typealias UiChange = (Ui) -> Unit
+
+class PinEntryCardController @Inject constructor(
+    private val userSession: UserSession,
+    private val passwordHasher: PasswordHasher
+) : ObservableTransformer<UiEvent, UiChange> {
+
+  override fun apply(events: Observable<UiEvent>): ObservableSource<UiChange> {
+    val replayedEvents = events.replay().refCount()
```

Comment:
We need to report analytics events too. Earlier, screens would report their events individually but this controller won't do that.

Reply:
Whoops, totally missed that. Will do this.